### PR TITLE
fix: kill nydus-image process with timeout

### PR DESCRIPTION
**Pull request by power-more**
Signed-off-by: zhaoshang <zhaoshangsjtu@linux.alibaba.com>

### CODE REVIEW

File: pkg/converter/tool/builder.go [hunk 1]
```diff
@@ -75,16 +79,17 @@ func Pack(option PackOption) error {
...
 
-	cmd := exec.Command(option.BuilderPath, args...)
+	cmd := exec.CommandContext(ctx, option.BuilderPath, args...)
 	cmd.Stdout = logger.Writer()
 	cmd.Stderr = logger.Writer()
 	cmd.Stdin = strings.NewReader(option.PrefetchPatterns)
 
 	if err := cmd.Run(); err != nil {
-		logrus.WithError(err).Errorf("fail to run %v %+v", option.BuilderPath, args)
+		logrus.WithError(err).Errorf("fail to run %v %+v, possibly due to timeout %v", option.BuilderPath, args, maxTimeOut)
```
Comment (imeoer) [inline pkg/converter/tool/builder.go:92]: Should check the error string `signal: killed` and then give `possibly due to timeout`.

File: pkg/converter/tool/builder.go [hunk 2]
```diff
@@ -7,13 +7,17 @@
...
+	"time"
 
 	"github.com/sirupsen/logrus"
 )
 
+const maxTimeOut = 2 * time.Hour
+
 var logger = logrus.WithField("module", "builder")
 
 type PackOption struct {
```
Comment (imeoer) [inline pkg/converter/tool/builder.go:23]: Can we define a `Timeout *time.Duration` field in `PackOption/MergeOption/UnpackOption`? `nil` means no timeout.

File: pkg/converter/tool/builder.go [hunk 3]
```diff
@@ -7,13 +7,17 @@
...
+	"context"
 	"fmt"
 	"os/exec"
 	"strings"
+	"time"
 
 	"github.com/sirupsen/logrus"
 )
 
+const maxTimeOut = 2 * time.Hour
```
Comment (imeoer) [inline pkg/converter/tool/builder.go:19]: `maxTimeout`

**Comment by changweige** [comment]
Could you please explain why `nydus-image` is hanging?

**Comment by power-more** [comment]
> Could you please explain why `nydus-image` is hanging?

Be prepared, after all, there is such a risk in theory.
In addition, we found in some business scenarios that in extreme cases (such as insufficient CPU resources and too large OCI layers), the conversion time of nydus-image process may be calculated in hours.

**Comment by power-more** [comment]
> Could you please explain why `nydus-image` is hanging?

update in git log : )

File: pkg/converter/tool/builder.go [hunk 4]
```diff
@@ -76,15 +82,25 @@ func Pack(option PackOption) error {
...
-	cmd := exec.Command(option.BuilderPath, args...)
+	cmd := exec.CommandContext(ctx, option.BuilderPath, args...)
 	cmd.Stdout = logger.Writer()
 	cmd.Stderr = logger.Writer()
 	cmd.Stdin = strings.NewReader(option.PrefetchPatterns)
 
 	if err := cmd.Run(); err != nil {
 		logrus.WithError(err).Errorf("fail to run %v %+v", option.BuilderPath, args)
+		if errdefs.IsSignalKilled(err) && option.Timeout != nil {
+			logrus.WithError(err).Errorf("the builder process is killed possibly due to timeout %v", *option.Timeout)
```
[the remark this answers is not in the document]
Comment (power-more) [reply pkg/converter/tool/builder.go]: fixed

File: pkg/errdefs/errors.go [hunk 5]
```diff
@@ -16,13 +16,19 @@ import (
...
 )
 
 // IsAlreadyExists returns true if the error is due to already exists
 func IsAlreadyExists(err error) bool {
 	return errors.Is(err, ErrAlreadyExists)
 }
 
+// IsSignalKilled returns true if the error is signal killed
+func IsSignalKilled(err error) bool {
+	return errors.Is(err, ErrSignalKilled)
```
[the remark this answers is not in the document]
Comment (power-more) [reply pkg/errdefs/errors.go:29]: no, fixing

[the remark this answers is not in the document]
Comment (power-more) [reply pkg/errdefs/errors.go]: fixed, use strings.Contains()

File: pkg/errdefs/errors.go [hunk 6]
```diff
@@ -9,20 +9,27 @@ package errdefs
...
 	"net"
+	"strings"
 	"syscall"
 
 	"github.com/pkg/errors"
 )
 
 var (
 	ErrAlreadyExists = errors.New("already exists")
+	SignalKilled     = "signal: killed"
```
Comment (imeoer) [inline pkg/errdefs/errors.go:20]: `const signalKilled = "signal: killed"` should be better.

Comment (power-more) [reply pkg/errdefs/errors.go]: fixed

**Review by imeoer — APPROVED** [review]
LGTM

**Review by changweige — APPROVED** [review]

**Pull request merged**